crates/sui-graphql-client/src/lib.rs

```diff
@@ -0,0 +1,568 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+#![doc = include_str!("../README.md")]
+
+pub mod graphql_types;
+
+use base64ct::Encoding;
+use graphql_types::{
+    ChainIdentifierQuery, CheckpointArgs, CheckpointId, CheckpointQuery, CoinMetadata,
+    CoinMetadataArgs, EpochSummaryArgs, EpochSummaryQuery, EventFilter, ObjectFilter, ObjectQuery,
+    ObjectQueryArgs, PageInfo, ProtocolConfigQuery, ProtocolConfigs, ProtocolVersionArgs,
+    ServiceConfig, ServiceConfigQuery, TransactionBlockArgs, TransactionBlockQuery, Uint53,
+};
+use sui_types::types::{
+    Address, CheckpointSequenceNumber, CheckpointSummary, Event, Object, SignedTransaction,
+};
+
+use anyhow::{ensure, Error, Result};
+use cynic::{serde, GraphQlResponse, Operation, QueryBuilder};
+use reqwest::Client;
+
+use crate::graphql_types::{
+    CoinMetadataQuery, EventsQuery, EventsQueryArgs, ObjectsQuery, ObjectsQueryArgs,
+};
+
+const MAINNET_HOST: &str = "https://sui-mainnet.mystenlabs.com/graphql";
+const TESTNET_HOST: &str = "https://sui-testnet.mystenlabs.com/graphql";
+const DEVNET_HOST: &str = "https://sui-devnet.mystenlabs.com/graphql";
+const DEFAULT_LOCAL_HOST: &str = "http://localhost:9125/graphql";
+
+/// An HTTP client based on reqwest
+struct ReqwestHttpClient {
+    inner: Client,
+}
+
+impl ReqwestHttpClient {
+    fn new() -> Self {
+        Self {
+            inner: Client::new(),
+        }
+    }
+
+    async fn post<
+        T: serde::de::DeserializeOwned + Send,
+        V: serde::Serialize + Send + std::marker::Sync,
+    >(
+        &self,
+        url: &str,
+        operation: &Operation<T, V>,
+    ) -> Result<GraphQlResponse<T>> {
+        let res = self
+            .inner
+            .post(url)
+            .json(&operation)
+            .send()
+            .await?
+            .json::<GraphQlResponse<T>>()
+            .await?;
+        Ok(res)
+    }
+}
+
+#[derive(Debug)]
+/// A page of items returned by the GraphQL server.
+pub struct Page<T> {
+    /// Information about the page, such as the cursor and whether there are more pages.
+    page_info: PageInfo,
+    /// The data returned by the server.
+    nodes: Vec<T>,
+}
+
+impl<T> Page<T> {
+    pub fn page_info(&self) -> &PageInfo {
+        &self.page_info
+    }
+
+    pub fn nodes(&self) -> &Vec<T> {
+        &self.nodes
+    }
+
+    fn new(page_info: PageInfo, nodes: Vec<T>) -> Self {
+        Self { page_info, nodes }
+    }
+}
+
+/// The SuiClient is a GraphQL client for interacting with the Sui blockchain.
+/// By default, it uses Reqwest as the HTTP client.
+pub struct SuiClient {
```

Comment:
Lets rename this to just `Client` and when we use the reqwest client we can just name it via its fully qualified type `reqwest::Client`. The additional "Sui" part of the name doesn't convey any additional information so it could be dropped. That and given namespacing we don't need to worry about this being named "Client" and something else being named "Client".

We could prepend it with GraphQlClient or some such, but even that isn't strictly necessary since the package itself is named sui-graphql-client.